lib/src/klio/message/pubsub_message_manager.py

```diff
@@ -109,38 +110,34 @@ def manage(self, message):
         Args:
             to_sleep(float): Seconds to sleep between checks.
         """
-        while not message.event.is_set():
-            self._maybe_extend(message)
-            time.sleep(self.manager_sleep)
-
-        else:
-            self.remove(message)
+        while True:
+            with MESSAGE_LOCK:
+                msg_is_active = message.kmsg_id in ENTITY_ID_TO_ACK_ID
+            if msg_is_active:
+                self._maybe_extend(message)
+                time.sleep(self.manager_sleep)
+            else:
+                self.remove(message)
+                break
 
     def heartbeat(self, message):
         """Continuously log heartbeats for in-progress messages.
 
         Args:
             to_sleep(float): Second to sleep between log messages.
         """
-        while not message.event.is_set():
-            self.hrt_logger.info(
-                f"Job is still processing {message.kmsg_id}..."
-            )
-            time.sleep(self.heartbeat_sleep)
-            # Sometimes the heartbeat worker still continues in the `while`
-            # loop even though the message event flag has been set (not sure
-            # how this can happen...).
-            # So we're just checking to see if the message we're looping
-            # over still exists in ENTITY_ID_TO_ACK_ID; and if not, to
-            # go ahead and remove it. The `remove` method should be
-            # idempotent.
-            if message.kmsg_id not in ENTITY_ID_TO_ACK_ID:
+        while True:
+            with MESSAGE_LOCK:
+                msg_is_active = message.kmsg_id in ENTITY_ID_TO_ACK_ID
+            if msg_is_active:
+                self.hrt_logger.info(
+                    f"Job is still processing {message.kmsg_id}..."
+                )
+                time.sleep(self.heartbeat_sleep)
+            else:
                 self.hrt_logger.debug(
-                    f"No record of message {message.kmsg_id}. Assuming the "
-                    "pipeline is done processing it and will try to ack the "
-                    "message..."
+                    f"Job is no longer processing {message.kmsg_id}."
                 )
-                self.remove(message)
```

Comment:
I removed this call to `remove` here, hoping it's not needed. But I'm going to test this new commit before moving forward with merging anything to make sure.